Port CreateAcceptance to the TCK

petraselmer opened the pull request:
Only merge this PR once https://github.com/neo4j/neo4j/pull/7215 has been merged

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,607 @@
+      """
+      MATCH (x:X)<-[:TYPE]-(y:Y)
+      RETURN x, y
+      """
```
I find that we want to have the query under test in the `When` step, and given that this file is meant to test `CREATE`, I'd say we want a query with `CREATE` in the `When` steps. It seems the `And` step has the create query here, and that the `When` step is meant to make assertions on the result. In #54 I added another `And` step that is meant to be run after the `When` step, to make possible such extra, query-based assertions after an updating query (I call these 'control queries').

Perhaps that's a better approach to this (and subsequent) scenarios? I do like to retain that `When` contains the query that the scenario wants to test.

petraselmer replied:
Totally agreed - will redo all to follow this format

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,607 @@
+      MATCH (x:Begin)
+      CREATE (x)-[:TYPE]->(:End)
+      """
+    When executing query:
```
See long comment above.

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,607 @@
+      """
+      CREATE (:A)<-[:R]-(:B)
+      """
+    When executing query:
```
See long comment above.

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,607 @@
+      """
+      CREATE (:A)-[:R]->(:B)-[:R]->(:C)
+      """
+    When executing query:
```
See long comment above.

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,607 @@
+      | +nodes  | 1 |
+
+  Scenario: Create a single node with properties
+    Given any graph
```
I don't think this is true. In the case of a populated uniqueness constraint, the query will not behave as specified here. I think we need `Given empty graph` here.

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,607 @@
+      | +properties | 1 |
+
+  Scenario: Creating a node with null properties should not return those properties
+    Given any graph
```
As above, `Given empty graph`.

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,607 @@
+      | +properties | 1 |
+
+  Scenario: Creating a relationship with null properties should not return those properties
+    Given any graph
```
As above, not applicable on any graph.

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,607 @@
+      | +properties    | 1 |
+
+  Scenario: Create a simple pattern
+    Given any graph
```
empty

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,607 @@
+      | +relationships | 1 |
+
+  Scenario: Create a self loop
+    Given any graph
```
empty

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,607 @@
+      | +labels        | 1 |
+
+  Scenario: Create a self loop using MATCH
+    Given any graph
```
And the last case, when there's a node property existence constraint on label `:R`, this won't work.

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,607 @@
+      """
+      CREATE (:A)<-[:R]-(:B)<-[:R]-(:C)
+      """
+    When executing query:
```
See long comment above.

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,607 @@
+      """
+      CREATE (:A)-[:R]->(:B)<-[:R]-(:C)
+      """
+    When executing query:
```
See long comment above.

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,607 @@
+      """
+      CREATE ()-[:R1]->()<-[:R2]-()-[:R3]->()
+      """
+    When executing query:
```
See long comment above.

Mats-SX commented on tck/features/LargeCreateQuery.feature:
```diff
@@ -0,0 +1,1360 @@
+      CREATE (Hugo:Person {name: 'Hugo Weaving', born: 1960})
+      CREATE (AndyW:Person {name: 'Andy Wachowski', born: 1967})
+      CREATE (LanaW:Person {name: 'Lana Wachowski', born: 1965})
+      CREATE (JoelS:Person {name: 'Joel Silver', born: 1952})
```
These variables aren't following Cypher style, as they are not camel case'd.

Mats-SX commented on tck/features/LargeCreateQuery.feature:
```diff
@@ -0,0 +1,1360 @@
+      | +labels        | 171 |
+
+  Scenario: Many CREATE clauses
+    Given any graph
```
empty

Mats-SX commented on tck/features/LargeCreateQuery.feature:
```diff
@@ -0,0 +1,1360 @@
+Feature: LargeCreateQuery
+
+  Scenario: Generate the movie graph correctly
+    Given any graph
```
empty

Mats-SX commented:
@petraselmer I stopped putting down line comments for the `any graph` vs `empty graph` condition. I think most of the scenarios need `empty graph`. I hadn't realised that creating a labelless node should always be possible, though, so that's good to know. But usually, when doing something that accesses the graph, `Given any graph` is too loose a condition to be able to state the outcome.

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+Feature: CreateAcceptance
+
+  Scenario: Create a single node
+    Given an empty graph
```
This could be `any graph`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+      | +nodes  | 1 |
+
+  Scenario: Create a single node with a single label
+    Given an empty graph
```
This could be `any graph`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+      | +labels | 1 |
+
+  Scenario: Create a single node with multiple labels
+    Given an empty graph
```
This could be `any graph`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+      | +nodes  | 1 |
+
+  Scenario: Create a single node with properties
+    Given an empty graph
```
This could be `any graph`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+    Given an empty graph
+    When executing query:
+      """
+      CREATE (n {id: 12, property: NULL})
```
`NULL` -> `null`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+      | +properties | 1 |
+
+  Scenario: Creating a node with null properties should not return those properties
+    Given an empty graph
```
This could be `any graph`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+      | +properties | 1 |
+
+  Scenario: Creating a relationship with null properties should not return those properties
+    Given an empty graph
```
This could be `any graph`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+    Given an empty graph
+    When executing query:
+      """
+      CREATE ()-[r:X {id: 12, property: NULL}]->()
```
`NULL` -> `null`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+      | +properties    | 1 |
+
+  Scenario: Create a simple pattern
+    Given an empty graph
```
This could be `any graph`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+      | +relationships | 1 |
+
+  Scenario: Create a self loop
+    Given an empty graph
```
This could be `any graph`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+      | +relationships | 1 |
+
+  Scenario: Create nodes and relationships
+    Given an empty graph
```
This could be `any graph`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+      | +relationships | 1 |
+
+  Scenario: Create a relationship with a property
+    Given an empty graph
```
This could be `any graph`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+      MATCH (x:X), (y:Y)
+      CREATE (x)<-[:TYPE]-(y)
+      """
+    And having executed:
```
This should be `When executing control query`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+      CREATE (:X)
+      CREATE (:Y)
+      """
+    When executing query:
```
This step should have its own `Then the result should be empty` and `And the side effects should be`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+      | +nodes | 4 |
+
+  Scenario: Create a relationship with a reversed direction - 1
+    Given an empty graph
```
This could be `any graph`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+    And no side effects
+
+  Scenario: Create a pattern with multiple hops - 1
+    Given an empty graph
```
This could be `any graph`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+    And no side effects
+
+  Scenario: Create a pattern with multiple hops in the reverse direction - 1
+    Given an empty graph
```
This could be `any graph`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+    And no side effects
+
+  Scenario: Create a pattern with multiple hops in varying directions - 1
+    Given an empty graph
```
This could be `any graph`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+    And no side effects
+
+  Scenario: Create a pattern with multiple hops with multiple types and varying directions - 1
+    Given an empty graph
```
This could be `any graph`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+      | +relationships | 2 |
+
+  Scenario: A bound node should be recognized after projection with WITH + WITH
+    Given an empty graph
```
This could be `any graph`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+      | +relationships | 1 |
+
+  Scenario: A bound node should be recognized after projection with WITH + UNWIND
+    Given an empty graph
```
This could be `any graph`

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,590 @@
+
+
+  Scenario: Fail when trying to create using an undirected relationship pattern
+    Given an empty graph
```
This could be `any graph`

Mats-SX commented on tck/features/LargeCreateQuery.feature:
```diff
@@ -0,0 +1,1361 @@
+Feature: LargeCreateQuery
+
+  Scenario: Generate the movie graph correctly
+    Given an empty graph
```
I _think_ this could be `any graph`.

Mats-SX commented on tck/features/LargeCreateQuery.feature:
```diff
@@ -0,0 +1,1361 @@
+      | +labels        | 171 |
+
+  Scenario: Many CREATE clauses
+    Given an empty graph
```
I _think_ this could be `any graph`.

Mats-SX commented:
@petraselmer I'm sorry for my above comment, and I don't agree with it anymore, for the most part. Now that we've cut the scope of openCypher to exclude constraints, creating is always possible. And if the scenario does not contain a control query with `MATCH`, then `any graph` is preferable.

Also, it seems we are not agreeing fully on the design of the control queries. You seem to have multiple scenarios to first assert on the `When` query, and then to assert on the control query. I prefer that we do this in one scenario, as per https://github.com/opencypher/openCypher/blob/master/tck/features/MergeIntoAcceptance.feature#L27. I think it makes sense to contain the assertion of the semantics in a single scenario, instead of splitting it up.

petraselmer commented:
@Mats-SX I have (hopefully) addressed your comments. No probs about the `any graph`. Also, the 'control query' construct is great - I didn't know about this before - and this gets rid of the frankly clumsy multiple scenario set up. I have now done this across the board...

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,523 @@
+      MATCH (m)
+      WITH n AS a, m AS b
+      CREATE (a)-[:T]->(b)
+      RETURN id(a) AS a, id(b) AS b
```
`id()` function not allowed; use a property instead.

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,523 @@
+      CREATE (a)-[:T]->(b)
+      WITH a AS x, b AS y
+      CREATE (x)-[:T]->(y)
+      RETURN id(x) AS x, id(y) AS y
```
`id()` function not allowed; use a property instead.

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,523 @@
+      CREATE (a)-[:T]->()
+      WITH a AS x
+      CREATE (x)-[:T]->()
+      RETURN id(x) AS x
```
`id()` function not allowed; use a property instead.

Mats-SX commented on tck/features/CreateAcceptance.feature:
```diff
@@ -0,0 +1,523 @@
+      MATCH (n)
+      WITH n AS a
+      CREATE (a)-[:T]->()
+      RETURN id(a) AS a
```
`id()` function not allowed; use a property instead.

Mats-SX commented:
@petraselmer Except for the `id()` function, this looks good. I find that usually the easiest way is to simply create an `id` property on the nodes, but it's also possible to work around by labelling and projecting the full nodes, or their labels using `labels()`.
